src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/ProcessedEnvelopeNotificationHandler.java

```diff
@@ -0,0 +1,158 @@
+package uk.gov.hmcts.reform.bulkscanprocessor.tasks;
+
+import com.fasterxml.jackson.core.JsonParseException;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.microsoft.azure.servicebus.ExceptionPhase;
+import com.microsoft.azure.servicebus.IMessage;
+import com.microsoft.azure.servicebus.IMessageHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import uk.gov.hmcts.reform.bulkscanprocessor.exceptions.EnvelopeNotFoundException;
+import uk.gov.hmcts.reform.bulkscanprocessor.exceptions.InvalidMessageException;
+import uk.gov.hmcts.reform.bulkscanprocessor.model.in.msg.ProcessedEnvelope;
+import uk.gov.hmcts.reform.bulkscanprocessor.services.EnvelopeFinaliserService;
+import uk.gov.hmcts.reform.bulkscanprocessor.services.servicebus.MessageAutoCompletor;
+
+import java.io.IOException;
+import java.util.concurrent.CompletableFuture;
+
+public class ProcessedEnvelopeNotificationHandler implements IMessageHandler {
+
+    private static final Logger log = LoggerFactory.getLogger(ProcessedEnvelopeNotificationHandler.class);
+
+    private final EnvelopeFinaliserService envelopeFinaliserService;
+    private final ObjectMapper objectMapper;
+    private final MessageAutoCompletor messageCompletor;
+
+    public ProcessedEnvelopeNotificationHandler(
+        EnvelopeFinaliserService envelopeFinaliserService,
+        ObjectMapper objectMapper,
+        MessageAutoCompletor messageCompletor
+    ) {
+        this.envelopeFinaliserService = envelopeFinaliserService;
+        this.objectMapper = objectMapper;
+        this.messageCompletor = messageCompletor;
+    }
+
+    @Override
+    public CompletableFuture<Void> onMessageAsync(IMessage message) {
+        return CompletableFuture
+            .supplyAsync(() -> tryProcessMessage(message))
+            .thenAcceptAsync(processingResult -> tryFinaliseMessage(message, processingResult))
+            .handleAsync((v, error) -> {
+                if (error != null) {
+                    log.error(
+                        "An error occurred when trying to handle 'processed envelope' message with ID {}",
+                        message.getMessageId()
+                    );
+                }
+
+                return null;
+            });
+    }
+
+    @Override
+    public void notifyException(Throwable throwable, ExceptionPhase exceptionPhase) {
+        log.error(
+            "An error occurred when handling processed envelope notification. Phase: {}",
+            exceptionPhase,
+            throwable
+        );
+    }
+
+    private void tryFinaliseMessage(IMessage message, MessageProcessingResult processingResult) {
+        try {
+            finaliseMessage(message, processingResult);
+        } catch (Exception e) {
```

Comment:
if you are catching the exception here - there's no need to handle in the future composition in main method above.

this whole method is pointless really. future is there to handle such misbehaviour

Reply:
The one in the main method catches anything that wasn't handled in any of the steps. There shouldn't be any such thing, but if it happens, we'd like to see it in logs. This catch, on the other hand (I've just changed it slightly, though) allows for providing an error message specific to this stage.